src/Aspire.Hosting/ApplicationModel/CommandsConfigurationExtensions.cs

```diff
@@ -117,4 +134,50 @@ internal static void AddLifeCycleCommands(this IResource resource)
         static bool IsStarting(string? state) => state is "Starting";
         static bool IsWaiting(string? state) => state is "Waiting" or "RuntimeUnhealthy";
     }
+
+    internal static async Task<ExecuteCommandResult> RunLifeCycleCommandAsync(
+        Func<ApplicationExecutor, CancellationToken, Task> action,
+        IServiceProvider serviceProvider,
+        string resourceName,
+        IEnumerable<string> expectedStates,
+        CancellationToken cancellationToken)
+    {
+        using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+        cts.CancelAfter(s_timeout);
+
+        var applicationExecutor = serviceProvider.GetRequiredService<ApplicationExecutor>();
+        var resourceNotificationService = serviceProvider.GetRequiredService<ResourceNotificationService>();
+
+        try
+        {
+            await action(applicationExecutor, cts.Token).ConfigureAwait(false);
+
+            await WaitForResourceAsync(resourceNotificationService, resourceName, expectedStates, cts.Token).ConfigureAwait(false);
+
+            return CommandResults.Success();
+        }
+        catch (OperationCanceledException) when (cts.Token.IsCancellationRequested)
+        {
+            return new ExecuteCommandResult
+            {
+                Success = false,
+                ErrorMessage = $"Timeout while waiting for '{resourceName}'."
+            };
+        }
+
+        static async Task<string> WaitForResourceAsync(ResourceNotificationService resourceNotificationService, string resourceId, IEnumerable<string> targetStates, CancellationToken cancellationToken = default)
+        {
+            await foreach (var resourceEvent in resourceNotificationService.WatchAsync(cancellationToken).ConfigureAwait(false))
+            {
+                if (string.Equals(resourceId, resourceEvent.ResourceId, StringComparisons.ResourceName)
+                    && resourceEvent.Snapshot.State?.Text is { Length: > 0 } statusText
+                    && targetStates.Contains(statusText, StringComparers.ResourceState))
+                {
+                    return statusText;
+                }
+            }
+
+            throw new OperationCanceledException($"The operation was cancelled before the resource reached one of the target states: [{string.Join(", ", targetStates)}]");
+        }
```

Comment:
You can delete this and use:

https://github.com/dotnet/aspire/blob/abba3d5a9e229823577ab679303afb9cce8d0069/src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs#L106

Reply:
Add a comment plz